File: sdk/node/src/resources/sandbox.ts
import type { HttpClient } from '../client.js';
import type { Sandbox, RunResult } from '../types.js';

export class SandboxResource {
  constructor(private readonly http: HttpClient) {}

  create(opts: { language: 'python' | 'nodejs' | 'bash'; timeout_secs?: number }): Promise<Sandbox> {
    return this.http.post('/v1/sandbox/create', opts);
  }

  get(sandboxId: string): Promise<Sandbox> {
    return this.http.get(`/v1/sandbox/${sandboxId}`);
  }

  run(sandboxId: string, opts: { code: string }): Promise<RunResult> {
    return this.http.post(`/v1/sandbox/${sandboxId}/run`, opts);
  }

  terminate(sandboxId: string): Promise<{ status: string }> {
    return this.http.delete(`/v1/sandbox/${sandboxId}`);
  }

  /** Convenience: create a sandbox, run code, terminate — returns the run result. */
  async runOnce(
    language: 'python' | 'nodejs' | 'bash',
    code: string,
    opts?: { timeout_secs?: number; waitMs?: number },
  ): Promise<RunResult> {
    const sb = await this.create({ language, timeout_secs: opts?.timeout_secs });

    // Poll until READY
    const waitMs = opts?.waitMs ?? 15_000;
    const deadline = Date.now() + waitMs;
    let current = sb;
    while (current.status === 'creating' && Date.now() < deadline) {
      await sleep(500);
      current = await this.get(sb.sandbox_id);
    }

    if (current.status !== 'ready') {
      await this.terminate(sb.sandbox_id).catch(() => {});
      throw new Error(`Sandbox failed to start (status=${current.status})`);
    }

    try {
      return await this.run(sb.sandbox_id, { code });
    } finally {
      await this.terminate(sb.sandbox_id).catch(() => {});
    }
  }
}

function sleep(ms: number) { return new Promise((r) => setTimeout(r, ms)); }

File: sdk/python/perpetual/resources/sandbox.py
from __future__ import annotations
import time
from typing import Literal
from ..http import HttpClient
from ..types import Sandbox, RunResult, PerpetualError


class SandboxResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def create(
        self,
        language: Literal["python", "nodejs", "bash"],
        timeout_secs: int = 300,
    ) -> Sandbox:
        data = self._http.post("/v1/sandbox/create", {"language": language, "timeout_secs": timeout_secs})
        return Sandbox(**data)

    def get(self, sandbox_id: str) -> Sandbox:
        data = self._http.get(f"/v1/sandbox/{sandbox_id}")
        return Sandbox(**data)

    def run(self, sandbox_id: str, code: str) -> RunResult:
        data = self._http.post(f"/v1/sandbox/{sandbox_id}/run", {"code": code})
        return RunResult(**data)

    def terminate(self, sandbox_id: str) -> dict:
        return self._http.delete(f"/v1/sandbox/{sandbox_id}")

    def run_once(
        self,
        language: Literal["python", "nodejs", "bash"],
        code: str,
        timeout_secs: int = 300,
        wait_secs: float = 15.0,
    ) -> RunResult:
        """Create sandbox, run code, terminate — all in one call."""
        sb = self.create(language, timeout_secs=timeout_secs)
        deadline = time.monotonic() + wait_secs

        while sb.status == "creating" and time.monotonic() < deadline:
            time.sleep(0.5)
            sb = self.get(sb.sandbox_id)

        if sb.status != "ready":
            self.terminate(sb.sandbox_id)
            raise PerpetualError(
                f"Sandbox failed to start (status={sb.status})", 0
            )

        try:
            return self.run(sb.sandbox_id, code)
        finally:
            self.terminate(sb.sandbox_id)
